Docker testing publish

BenjaminDev opened the pull request:
# Pull Request

## Description
Publish to Docker Hub so it's easy for people to try out the library.

Changes:
- added a new workflow that builds the image, runs the test suit and (in theory - credentials will be needed) pushes the image to docker hub.
    
Notes: Dev dependencies are not baked into the image but only used to run tests. 

Fixes OpenMined#163
## Affected Dependencies
None

## Type of Change
Please mark options that are relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation (non-breaking change which adds documentation)
- [x]  Improvement (non-breaking change that improves the performance or reliability of existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How has this been tested?
- Github action is successful. 
- To reporduce: fork and push a change to `dev`. Note: make sure your github repo settings enables actions to run. Seems my defaults for forks in not to run the actions.  


## Checklist
- [ ] I did follow the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md)
- [x] I did follow the [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have added tests for my changes

## Additional Context
Questions to conclude this PR
- When (which branches) should trigger this action? 
- What tagging convention should we use? (So far just latest and git sha) 
- What image naming should we use?
- We'll need to add a github secret with docker hub creds allow pushing once the rest has  been done.
- Should the tests we run on these images be against all supported python versions? I am unsure how the current tox setup is working but I'm sure we could run that. Depending on the intended use of these images I guess. 


I'm super happy to complete these but someone who knows those answers can pop them in and take over if they would like.

BenjaminDev committed:
added github workflow to publish the docker image. Runs test suit as a sanity check.

BenjaminDev committed:
Update update-docker-image.yml

BenjaminDev committed:
Merge pull request #2 from BenjaminDev/bg/163-publish-docker-image

Bg/163 publish docker image

BenjaminDev committed:
Update update-docker-image.yml

BenjaminDev committed:
Update update-docker-image.yml

BenjaminDev committed:
Update update-docker-image.yml

codeboy5 commented:
I think we can have the publish workflow to run only a specific branch or tag, and the build needs to run on every PR to check the Docker-file

BenjaminDev committed:
added builds for all python versions.

BenjaminDev committed:
syntax

BenjaminDev committed:
formatting remove spaces.

BenjaminDev committed:
testing

BenjaminDev committed:
shuffle the image name concatenation.

BenjaminDev committed:
need to references from steps level

BenjaminDev committed:
need to references from steps level

BenjaminDev committed:
Fix docker args. hyphens are illegal it seems.

chinmayshah99 commented on .github/workflows/update-docker-image.yml:
```diff
@@ -0,0 +1,50 @@
+      - name: Run tests
+        run:
+          | # Not sure if this is what is desired. If a test fails the whole build fails - can mark flaky tests if needed?
+          docker run ${{steps.build_name.outputs.image_name}}:latest /bin/bash -c  "pipenv install --dev && pipenv run pytest ./tests"
```
This can be a separate workflow

BenjaminDev replied:
I think it would be nicer to run the tests against the docker images before pushing them out the door. 
It won't add much to the total time taken to run the action. And the final shipped artifact will be identical.

chinmayshah99 replied:
We want to run docker tests one each commit. 
But publishing it to dockerhub is not something we want to do for each commit

chinmayshah99 commented on .github/workflows/update-docker-image.yml:
```diff
@@ -0,0 +1,50 @@
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: [3.6, 3.7, 3.8]
```
Can we add python 3.5 too?

zizu-research replied:
Sure. I'd forgotten to check that.

BenjaminDev replied:
Sure. I forgot to check that. Thanks

chinmayshah99 commented on Dockerfile:
```diff
@@ -1,5 +1,6 @@
+ARG python_version
+FROM python:${python_version}-slim-buster
 
 # Set environment variables.
```
Can you resolve the merge conflict?
Our latest dockerfile deals with this

chinmayshah99 commented on Dockerfile:
```diff
@@ -50,7 +51,7 @@ RUN \
     rm -rf dist/ && \
     pipenv run python3 setup.py bdist_wheel && \
-    pip3 install dist/*.whl
+    pipenv install dist/*.whl
```
If we do this. then when we open the bash, would the pipenv env open or the default python env?
The question is. if I type 
```
python3
>> import pydp as dp
```
Would this work?

BenjaminDev replied:
I believe we should install the final package into a virtual environment. If you don't want the user to have to run `pipenv shell` and to simply run `python` then I think this would work.
```
ENV PIPENV_VENV_IN_PROJECT=true 
# This `activates` the virtual env
ENV VIRTUAL_ENV=$PROJECT_DIR/.venv
ENV PATH="$VIRTUAL_ENV/bin:$PATH"
```
(edit: tested and it works)
However, to just install it outside a virtual env is not ideal and could have unforeseen issues. We could always add a line to the .bashrc file to put the user into the relevant environment?
Perhaps I need to full understand the use of these images and I can do my best to make them work as expected.

My current understanding of what we want out of this is: 
-  Publish docker images to Docker Hub so it's easy for people to try out the library.

chinmayshah99 replied:
Yeah. this is what we are aiming at!

BenjaminDev committed:
WIP. need to version upstream deps like thet base dp lib from google. Otherwise all our stuff will break unexpectedly. Better to formally pull in deps. However, I see  doesn't have any releases yet.

BenjaminDev committed:
Merge branch 'dev' of https://github.com/OpenMined/PyDP into dev

BenjaminDev committed:
Clone submodules as well so we can peg to a SHA and not break unexpectedly. Include python 3.5 in build images

BenjaminDev committed:
remove 3rd party from ignore and clone the submodule instead.

BenjaminDev committed:
Remove cloning 3rd party deps as it was not pegged to a SHA or version. Using git submodule to ensure we pull in the correct version of 3rd code.

BenjaminDev committed:
explicity stated the pipenv place the venv inside the project dir. Set env vars to  the appropriate virtual env so users can simply run

BenjaminDev committed:
updated python version arg to upper case

BenjaminDev committed:
remove debugging

BenjaminDev committed:
debugging GITHUB_SHA

BenjaminDev committed:
debugging GITHUB_SHA

BenjaminDev committed:
removed debugging

BenjaminDev commented:
Remaining items:

1. Are the image names and tags correct? Currently  openmined/pydp-{branch}-{python-version} Is this consistent across the organization?
```
REPOSITORY                                           TAG                                
openmined/pydp-dev-3.6                               eb3f50ce77c50886051309cdcf7708e7d40ca94a  
openmined/pydp-dev-3.6                               latest   
```
2. Un-comment the last stage to push the images to docker hub and add the credentials to github secrets. 
3. Remove verbose comments and scratch marks in Dockerfile. I left it to assist the reviewer.

Notes: I made changes the the Dockerfile as it cloned the `thirdparty` code which was not pegged to a version or sha. Instead I copy this in by cloning the submodule (same as a local build). l don't see any issues with that but just wanted to mention that incase I have missed something

chinmayshah99 commented on Dockerfile:
```diff
@@ -45,23 +45,37 @@ WORKDIR ${PROJECT_DIR}
+#     cd third_party && \
+#     git clone https://github.com/google/differential-privacy.git
+
+# Or we just copy it in and ensure the CI/CD clone the repo and the submodule.
```
The issue with this is when we do run this in remote environment,  `https://github.com/google/differential-privacy.git` is going to be a submodule which is not going to be cloned. Thus this will break your system. 
You can verify this by simply cloning the project and trying to build the dokcerfile. 
So what we can do here is probably stick to mentioning the sha.

zizu-research replied:
Please see the build it all passes as I pull all submodules 
In github actions one just need to do:
```
steps:
      - uses: actions/checkout@v2
        with:
          submodules: true
```
"So what we can do here is probably stick to mentioning the sha." 
That was the alternative suggested but it's very hidden I believe it'll break as people **will** forget. 

FYI:
If you want to clone a repo with it's submodules (Which is what I do on my local build) simply use:
```
git clone --recurse-submodules https://github.com/OpenMined/PyDP.git
``` 
This will correctly clone the repo at the sha of upstream as shown in the repo.

BenjaminDev replied:
That is not a problem as we clone the submodules in the github action already 
See the github action file.
```
steps:
      - uses: actions/checkout@v2
        with:
          submodules: true
```

As for the local builds one just needs to clone the submodules using `--recurse-submodules `

FYI:
For running things locally or in any other checkouts one just needs to clone as follows. 
```
git clone --recurse-submodules https://github.com/OpenMined/PyDP.git
```
This will correctly clone the upstream dependencies (`third_party`) at the correct sha. 

"So what we can do here is probably stick to mentioning the sha."
This is an alternative as suggested however, it is a very hidden config an **will** be forgotten and things will break (or not break when they should) as that sha will become out of sync with the one mentioned in the submodule. I'm in favor of letting git handle as much as possible.

However the remaining issues do still remain.
https://github.com/OpenMined/PyDP/pull/176#issuecomment-641100600

chinmayshah99 replied:
Oh, I didn't notice the submodule in actions.  
For running locally, we already have scripts that take cares of this stuff.

chinmayshah99 replied:
> 1. Are the image names and tags correct? Currently  openmined/pydp-{branch}-{python-version} Is this consistent across the organization?

This looks good. 

> 1. Un-comment the last stage to push the images to docker hub and add the credentials to github secrets.
> 2. Remove verbose comments and scratch marks in Dockerfile. I left it to assist the reviewer.

I'll make these changes while adding the credentials

chinmayshah99 committed:
Merge branch 'dev' into dev

chinmayshah99 committed:
added docker publish stuff

chinmayshah99 committed:
added third_party yet again to docker file

It made sense, as people using docker for development already have third-party folder cloned.

chinmayshah99 committed:
disabled docker hub push on every release

chinmayshah99 commented:
Hey @BenjaminDev thanks for the PR. Why don't you join us on Slack?

chinmayshah99 reviewed (approved):
LGTM!
